Code/GraphMol/MonomerMol/MMToAtomistic.cpp

```diff
@@ -0,0 +1,319 @@
+{
+    // in form RX-RY, returns {X, Y}
+    auto dash = linkage.find('-');
+    if (dash == std::string::npos) {
+        std::string error_msg = "Invalid linkage format: " + linkage;
+        throw std::runtime_error(error_msg);
+    }
+    return {std::stoi(linkage.substr(1, dash - 1)),
+            std::stoi(linkage.substr(dash + 2))};
+}
+
+void fill_attachment_point_map(const RDKit::ROMol& new_monomer,
+                               AttachmentMap& attachment_points,
+                               unsigned int residue_num,
+                               unsigned int old_mol_size)
+{
+    for (const auto& atom : new_monomer.atoms()) {
+        unsigned int map_num;
+        if (atom->getPropIfPresent(RDKit::common_properties::molAtomMapNumber,
+                                   map_num)) {
+            for (auto bnd : new_monomer.atomBonds(atom)) {
+                // Should be exactly one iteration -- an attachment point can
+                // only have one neighbor
+                if (attachment_points.find({residue_num, map_num}) !=
+                    attachment_points.end()) {
+                    std::string error_msg = "Invalid attachment point at index " +
+                                            std::to_string(atom->getIdx());
+                    throw std::runtime_error(error_msg);
+                }
+                auto atom_to_bond_to =
+                    old_mol_size + bnd->getOtherAtomIdx(atom->getIdx());
+                auto atom_to_remove = old_mol_size + atom->getIdx();
+                attachment_points[{residue_num, map_num}] =
+                    std::make_pair(atom_to_bond_to, atom_to_remove);
+            }
+        }
+    }
+}
+
+void set_pdb_info(RDKit::RWMol& new_monomer, const std::string& monomer_label,
+                  unsigned int residue_number, char chain_id,
+                  ChainType chain_type)
+{
+    std::string residue_name =
+        (chain_type == ChainType::PEPTIDE) ? "UNK" : "UNL";
+    if (three_character_codes.find(monomer_label) !=
+        three_character_codes.end()) {
+        residue_name = three_character_codes.at(monomer_label);
+    }
+
+    // Set PDB residue info on the atoms
+    for (auto atom : new_monomer.atoms()) {
+        auto* res_info = new RDKit::AtomPDBResidueInfo();
+
+        std::string chain_id_str(1, chain_id);
+        res_info->setChainId(chain_id_str);
+        res_info->setResidueNumber(residue_number);
+        res_info->setResidueName(residue_name);
+        // to be consistent with the rdkit adapter and RDKit's own PDB writer
+        res_info->setInsertionCode(" ");
+
+        std::string pdb_name;
+        if (atom->getPropIfPresent(ATOM_PDB_NAME_PROP, pdb_name)) {
+            res_info->setName(pdb_name);
+            // Don't keep property on output molecule
+            atom->clearProp(ATOM_PDB_NAME_PROP);
+        }
+
+        atom->setMonomerInfo(res_info);
+    }
+}
+
+ChainType get_chain_type(std::string_view polymer_id)
```

Comment:
we should remove this

Reply:
its currently used below -- maybe rename to `chain_name_to_type`?

Reply:
my bad :)